flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderRetrievalService.java

```diff
@@ -0,0 +1,178 @@
+ */
+class KubernetesLeaderRetrievalService implements LeaderRetrievalService {
+
+	private static final Logger LOG = LoggerFactory.getLogger(KubernetesLeaderRetrievalService.class);
+
+	private final Object lock = new Object();
+
+	private final FlinkKubeClient kubeClient;
+
+	private final String configMapName;
+
+	@GuardedBy("lock")
+	private volatile String lastLeaderAddress;
+
+	@GuardedBy("lock")
+	private volatile UUID lastLeaderSessionID;
+
+	@GuardedBy("lock")
+	private volatile LeaderRetrievalListener leaderListener;
+
+	@GuardedBy("lock")
+	private volatile boolean running;
+
+	private KubernetesWatch kubernetesWatch;
+
+	KubernetesLeaderRetrievalService(FlinkKubeClient kubeClient, String configMapName) {
+		this.kubeClient = checkNotNull(kubeClient, "Kubernetes client should not be null.");
+		this.configMapName = checkNotNull(configMapName, "ConfigMap name should not be null.");
+
+		this.leaderListener = null;
+		this.lastLeaderAddress = null;
+		this.lastLeaderSessionID = null;
+
+		running = false;
+	}
+
+	@Override
+	public void start(LeaderRetrievalListener listener) {
+		checkNotNull(listener, "Listener must not be null.");
+		Preconditions.checkState(leaderListener == null, "KubernetesLeaderRetrievalService can " +
+			"only be started once.");
+
+		LOG.info("Starting {}.", this);
+
+		synchronized (lock) {
+			running = true;
+			leaderListener = listener;
+			kubernetesWatch = kubeClient.watchConfigMaps(configMapName, new ConfigMapCallbackHandlerImpl());
+		}
+	}
+
+	@Override
+	public void stop() {
+		LOG.info("Stopping {}.", this);
+
+		synchronized (lock) {
+			if (!running) {
+				return;
+			}
+			running = false;
+			if (kubernetesWatch != null) {
+				kubernetesWatch.close();
+			}
+		}
+	}
+
+	@Override
+	public String toString() {
+		return "KubernetesLeaderRetrievalService{configMapName='" + configMapName + "'}";
+	}
+
+	private class ConfigMapCallbackHandlerImpl implements FlinkKubeClient.WatchCallbackHandler<KubernetesConfigMap> {
+
+		@Override
+		public void onAdded(List<KubernetesConfigMap> configMaps) {
+			handleEvent(configMaps);
+		}
+
+		@Override
+		public void onModified(List<KubernetesConfigMap> configMaps) {
+			handleEvent(configMaps);
+		}
+
+		@Override
+		public void onDeleted(List<KubernetesConfigMap> configMaps) {
+			// Nothing to do since a new ConfigMap will be created if it is deleted externally.
+		}
+
+		@Override
+		public void onError(List<KubernetesConfigMap> configMaps) {
+			leaderListener.handleError(new Exception("Error while watching the ConfigMap " + configMapName));
+		}
+
+		@Override
+		public void handleFatalError(Throwable throwable) {
+			leaderListener.handleError(
+				new Exception("Fatal error while watching the ConfigMap " + configMapName, throwable));
+		}
+
+		private void handleEvent(List<KubernetesConfigMap> configMaps) {
+			synchronized (lock) {
+				if (running) {
+					configMaps.forEach(e -> {
+						if (e.getName().equals(configMapName)) {
```

Comment:
Should we add an assertion on the config map name?

Reply:
Yes.